README.md
<p align="center">
 <img src="https://github.com/Boomshicleafaunda/Support_BaseplateRules/blob/master/title.png" height=64>
</p>

<p align="center"><sup><i>Adds baseplate alignment and adjacency rules for building</i></sup></p>

<p align="center"><strong>Release:</strong> 1.0.2</p>
<p align="center"><strong>Download:</strong>
 <a href="https://github.com/Boomshicleafaunda/Support_BaseplateRules/archive/v1.0.2.zip">GitHub</a> | <a href="https://blocklandglass.com/addons/download.php?id=1032">Blockland Glass</a>
</p>

## About Baseplate Rules

Baseplate Rules provide Alignment and Adjacency rules for building. When Baseplate Rules are enabled, all players will be required to start all of their builds on a baseplate. When Baseplate Alignment is enabled, the starting baseplate must be aligned to a configurable grid. When Baseplate Adjacency is enabled, the starting baseplate must be placed adjacent (directly next to) another baseplate.

## Preferences for Baseplate Rules

Baseplate Rules are completely configurable. Here are all of the preferences:

### Baseplate Preferences

#### Baseplate Required

When enabled, players are required to start all of their builds on a baseplate.

> **Note:** You can use the `/baseplaterules enable` command require baseplates. Similarly, you can use the `/baseplaterules disable` command to no longer require baseplates. By default, baseplate rules are enabled.

#### Baseplate Type

It can be configured such that not all baseplates from the "Baseplates" category can be used as starting baseplates. Using this preference, you can configure the starting baseplate to be one of three types:

 - **Any:** Any brick from the baseplate category.
 - **Plate:** Any 1/3x tall plate brick from the baseplate category.
 - **Plain:** Any brick from the Plain baseplate subcategory.
 
 > **Note:** You can use the `/baseplaterules type set <type>` command to set the required baseplate type. By default, the required baseplate type is set to "Any".

### Adjacency Preferences

#### Use Adjacency

When enabled, players are required to place their starting baseplate adjacent to another baseplate.

> **Note:** You can use the `/baseplaterules adjacency enable` command to enable adjacency rules. Similarly, you can use the `/baseplaterules adjacency disable` command to disable adjacency rules. By default, adjacency rules are enabled.

#### Include Diagonals

When enabled, baseplates diagonally positioned are considered adjacent.

> **Note:** You can use the `/baseplaterules adjacency diagonals enable` command to allow diagonal adjacency. Similarly, you can use the `/baseplaterules adjacency diagonals disable` command to no longer allow diagonal adjacency. By default, diagonal adjacency is not allowed.

### Alignment Preferences

#### Use Alignment

When enabled, players are required to align their baseplates to a configurable grid.

> **Note:** You can use the `/baseplaterules alignment enable` command to enable alignment rules. Similarly, you can use the `/baseplaterules alignment disable` command to disable alignment rules. By default, alignment rules are enabled.

#### Auto Alignment

When enabled, players attempting to place a misaligned baseplate will have their ghost brick automatically realigned for them.

> **Note:** You can use the `/baseplaterules alignment auto enable` command to enable auto alignment. Similarly, you can use the `/baseplaterules alignment auto disable` command to disable auto alignment. By default, auto alignment is enabled.

#### Grid Cell Size

The alignment grid is configurable in such a way that allows you to control the grid cell size.

> **Note:** You can use the `/baseplaterules alignment grid set <X> <Y>` command to update the grid cell size. Both the `<X>` and `<Y>` values must be valid integers between 1 and 1024. By default, the grid cell size is set to 16x16.

#### Grid Offset

The alignment grid is configurable in such a way that allows you to control the grid offset. This should really only be used when trying to snap the grid to a preexisting build or terrain-based obstable. In an flat freebuild world, adjusting the grid offset is unnecessary.

> **Note:** You can use the `/baseplaterules alignment offset set <X> <Y>` command to update the grid offset. Both the `<X>` and `<Y>` values must be valid integers between 0 and 1024. By default, the grid offset is set to 0x0.


CHANGELOG.md
# Release Notes

## v1.0.1 (2019-01-17)

### Added
 - Support for New Duplicator

### Fixed
 - Adjacency checks when alignment is not enabled
 - Bricks no longer count as adjacent when in their death animation
 - Various quirks with diagonal adjacency

## v1.0.2 (2019-01-21)

### Fixed
 - Baseplate Rules are now ignored while loading bricks

Scripts/package.cs
//* Description *//
// Title: Package
// Author: Boom (9740)
// Defines the package for this add-on.

// Disable Existing Package
if(isPackage(Support_BaseplateAlignment_Package)) {
	deactivatePackage(Support_BaseplateAlignment_Package);
}

//* Package *//
package Support_BaseplateAlignment_Package
{
	// /**
	//  * Triggered when a client plants a brick.
	//  *
	//  * @param  {GameConnection}  %client  The client attempting to plant a brick.
	//  *
	//  * @return {void}
	//  */
	function serverCmdPlantBrick(%client)
	{
		// Perform the baseplate plant brick check
		if(!%client.doBaseplatePlantBrickCheck()) {
			return;
		}

		// Use the default functionality
		return Parent::serverCmdPlantBrick(%client);
	}

	// Certain mods will plant bricks without calling the plant brick
	// server command. Just to cover our bases, we're going to use
	// overrides for both "/plantbrick" and fxDTSBrick.plant().

	// /**
	//  * Attempts to plant the calling brick.
	//  *
	//  * @param  {fxDTSBrick}  %brick  The calling brick.
	//  *
	//  * @return {integer}
	//  */
	function fxDTSBrick::plant(%brick)
	{
		// Determine the client
		%client = %brick.client;

		// If the brick has a client, perform the baseplate plant brick check
		if(isObject(%client) && !%client.doBaseplatePlantBrickCheck(%brick)) {
			return 4; // Unstable
		}

		// If there is no client, still perform the baseplate plant brick check
		if(!isObject(%client) && !%brick.doBaseplateCheck()) {
			return 4; // Unstable
		}

		// Use the default functionality
		return Parent::plant(%brick);
	}

	// If you are a cheeky bastard, then you may have noticed that a
	// brick in its death animation still counts as adjaceny. Even
	// though the brick can be killed later on, it's not clean.

	// /**
	//  * Triggered when the calling brick is being killed.
	//  *
	//  * @param  {fxDTSBrick}  %brick  The calling brick.
	//  *
	//  * @return {integer}
	//  */
	function fxDTSBrick::onDeath(%brick)
	{
		// Mark the brick as dead
		%brick.isDead = true;

		// Use the default functionality
		return Parent::onDeath(%brick);
	}

	// /**
	//  * Triggered for each tick of a save file being loaded.
	//  *
	//  * @return {void}
	//  */
	function ServerLoadSaveFile_Tick()
	{
		// If baseplate rules are disabled, carry on
		if(!$Support::Baseplate::Required) {
			return Parent::ServerLoadSaveFile_Tick();
		}

		// When baseplate rules are enabled, certain bricks may get
		// loaded that violate baseplate rules. However, we can't
		// assume that the bricks are supposed to follow rules.

		// Temporarily disable baseplate rules
		$Support::Baseplate::Required = 0;

		// Call the parent function
		%result = Parent::ServerLoadSaveFile_Tick();

		// Enable baseplate rules
		$Support::Baseplate::Required = 1;

		// Return the result
		return %result;
	}

};

// Activate the package
activatePackage(Support_BaseplateAlignment_Package);

Scripts/adjacency.cs
//* Description *//
// Title: Adjacency
// Author: Boom (9740)
// Defines the adjacency portion of this add-on.

// /**
//  * Triggered when a brick is checked for baseplate neighbors.
//  *
//  * @param  {fxDTSBrick}  %brick   The brick performing the neighbor check.
//  *
//  * @return {boolean}
//  */
function fxDTSBrick::onAdjacencyCheck(%brick)
{
	// This method would be a good spot for a package to add any
	// overrides for admins or something if you wanted to have
	// scenarios where you could ignore the adjacency rules.

	// One quirk of adjacency is that if there aren't any bricks
	// planted, there's nothing to be adjacent to. If this is
	// the first brick, we should ignore adjacency rules.

	// If this is the first brick, ignore adjacency rules
	if(%brick.isFirstBrick()) {
		return false;
	}

	// Return whether or not to perform the neighbor check
	return true;
}

// /**
//  * Triggered when a baseplate passes a neighbor check.
//  *
//  * @param  {fxDTSBrick}  %brick  The brick that passed the neighbor check.
//  *
//  * @return {void}
//  */
function fxDTSBrick::onAdjacencyPass(%brick)
{
	// This method purely exists for package hooks, and serves no
	// purpose in this mod's default functionality. However, an
	// external add-on may choose to add a listener to this.
}

// /**
//  * Triggered when a baseplate fails a neighbor check.
//  *
//  * @param  {fxDTSBrick}  %brick  The brick that failed the neighbor check.
//  *
//  * @return {void}
//  */
function fxDTSBrick::onAdjacencyFail(%brick)
{
	// This method purely exists for package hooks, and serves no
	// purpose in this mod's default functionality. However, an
	// external add-on may choose to add a listener to this.
}

// /**
//  * Performs a neighbor check on the calling brick.
//  *
//  * @param  {fxDTSBrick}  %brick  The calling brick.
//  *
//  * @return {boolean}
//  */
function fxDTSBrick::doAdjacencyCheck(%brick)
{
	// Call the neighbor check to make sure we should do it
	%result = %brick.onAdjacencyCheck();

	// If the result is strictly false, then do not perform the check
	if(%result $= 0) {
		return true;
	}

	// Perform the neighbor check
	%success = %brick.hasAdjacentBaseplate();

	// Call the success trigger
	if(%success) {
		%brick.onAdjacencyPass();
	} else {
		%brick.onAdjacencyFail();
	}

	// Return whether or not the check was successful
	return %success;
}

// /**
//  * Returns whether or not the calling brick has a baseplate neighbor.
//  *
//  * @param  {fxDTSBrick}  %brick  The calling brick.
//  *
//  * @return {boolean}
//  */
function fxDTSBrick::hasAdjacentBaseplate(%brick)
{
	// Before we can perform an adjacency check, there's a lot of info
	// that we need to know. We'll pull everything together ahead of
	// time, and perform the check near the end of this function.

	// Determine the alignment of the brick
	%alignment = %brick.getAlignment();

	%alignmentX = getWord(%alignment, 0);
	%alignmentY = getWord(%alignment, 1);
	%alignmentZ = getWord(%alignment, 2);

	// Determine the center of the brick
	%brickPos = getWords(%brick.getTransform(), 0, 2);

	// Determine the grid size of the brick
	%brickCellSize = %brick.getAlignmentGridSize();

	%brickCellWidth = getWord(%brickCellSize, 0);
	%brickCellLength = getWord(%brickCellSize, 1);

	// Determine the grid cell size
	%gridCellWidth = $Support::Baseplate::Alignment::Width;
	%gridCellLength = $Support::Baseplate::Alignment::Length;

	// If baseplate alignment isn't actually enabled, then we can't rely
	// on it. We can however, assume the grid to be comprised of 1x1
	// cells, which will allow us to check adjacency at each stud.

	// Check if alignment is disabled
	if(!$Support::Baseplate::Alignment::Enabled) {

		// Use a 1x1 grid
		%gridCellWidth = 1;
		%gridCellLength = 1;

	}

	// The global position of bricks is not as neat as one would expect.
	// Every 1 meter counts as 2 blocks laterally. Because of this,
	// we need to scale down our grid to match the global size.

	// Scale the grid
	%gridCellWidth /= 2;
	%gridCellLength /= 2;

	// The starting positions will be the center alignment offset by half
	// of the grid (plus a little give) in cell size units. This should
	// end up being the top left adjacent cell where a brick may be.

	// Determine the offset
	%offsetX = (((%brickCellWidth / 2) + 0.5) * %gridCellWidth);
	%offsetY = (((%brickCellLength / 2) + 0.5) * %gridCellLength);

	// Determine the starting positions for checking (left and top)
	%startX = %alignmentX - %offsetX;
	%startY = %alignmentY - %offsetY;

	// Determine the ending positions for checking (right and bottom)
	%endX = %alignmentX + %offsetX;
	%endY = %alignmentY + %offsetY;

	// To determine the bottom of the brick, we must first calculate the
	// height in meters. Every 1 meter counts as 5 plates vertically.
	// We then need half that value to go from center to bottom.

	// Use the bottom of the brick the current height position
	%currentZ = getWord(%brickPos, 2) - %brick.getDatablock().brickSizeZ / 10;

	// The absolute bottom of the brick is not actually what we want. A
	// raycast would skim just slightly under the brick and be unable
	// to find anything. We should go up by half a plate instead.

	// Increase the current height position by half a plate
	%currentZ += 0.05;

	// We are going to raycast from the bottom origin of the brick to a
	// possible location of an adjacent brick. Before we can do that,
	// we must of course calculate the origin we should be using.

	// Determine the raycast origin
	%origin = getWord(%brickPos, 0) SPC getWord(%brickPos, 1) SPC %currentZ;

	// Iterate through the top alignment
	for(%currentX = %startX; %currentX <= %endX; %currentX += %gridCellWidth) {

		// Exclude diagonals when we're told to
		if(!$Support::Baseplate::Adjacency::IncludeDiagonals && (%currentX == %startX || %currentX == %endX)) {
			continue;
		}

		// Determine the current position
		%currentPos = %currentX SPC %startY SPC %currentZ;

		// Check for an adjacent baseplate
		if(%brick.checkForAdjacentBaseplate(%origin, %currentPos)) {
			return true;
		}

	}

	// Iterate through the bottom alignment
	for(%currentX = %startX; %currentX <= %endX; %currentX += %gridCellWidth) {

		// Exclude diagonals when we're told to
		if(!$Support::Baseplate::Adjacency::IncludeDiagonals && (%currentX == %startX || %currentX == %endX)) {
			continue;
		}

		// Determine the current position
		%currentPos = %currentX SPC %endY SPC %currentZ;

		// Check for an adjacent baseplate
		if(%brick.checkForAdjacentBaseplate(%origin, %currentPos)) {
			return true;
		}

	}

	// Iterate through the left alignment
	for(%currentY = %startY; %currentY <= %endY; %currentY += %gridCellLength) {

		// Exclude diagonals when we're told to
		if(!$Support::Baseplate::Adjacency::IncludeDiagonals && (%currentY == %startY || %currentY == %endY)) {
			continue;
		}

		// Determine the current position
		%currentPos = %startX SPC %currentY SPC %currentZ;

		// Check for an adjacent baseplate
		if(%brick.checkForAdjacentBaseplate(%origin, %currentPos)) {
			return true;
		}

	}

	// Iterate through the right alignment
	for(%currentY = %startY; %currentY <= %endY; %currentY += %gridCellLength) {

		// Exclude diagonals when we're told to
		if(!$Support::Baseplate::Adjacency::IncludeDiagonals && (%currentY == %startY || %currentY == %endY)) {
			continue;
		}

		// Determine the current position
		%currentPos = %endX SPC %currentY SPC %currentZ;

		// Check for an adjacent baseplate
		if(%brick.checkForAdjacentBaseplate(%origin, %currentPos)) {
			return true;
		}

	}

	// No adjacent baseplates were found
	return false;
}

// /**`
//  * Returns whether or not an adjancent baseplate exists at the specified position.
//  *
//  * @param  {fxDTSBrick}  %brick     The calling brick.
//  * @param  {string}      %origin    The origin (bottom center) of the brick.
//  * @param  {string}      %position  The position to check for an adjacent baseplate.
//  *
//  * @return {boolean}
//  */
function fxDTSBrick::checkForAdjacentBaseplate(%brick, %origin, %position)
{
	// Perform a raycast from the center of the brick to the current position
	%object = firstWord(containerRaycast(%origin, %position, $TypeMasks::FxBrickAlwaysObjectType, %brick));

	// Sometimes raycasts can be a bit tricky. If we don't find anything
	// we can use, then we should try a container search at the place
	// where we think a brick should be. You never know with this.

	// If an object was not found, try a different type of search
	if(!isObject(%object)) {
		%object = containerFindFirst($TypeMasks::FxBrickAlwaysObjectType, %position, 0.125, 0.125, 0.05);
	}

	// If an object was still not found, then no baseplate exists
	if(!isObject(%object)) {
		return false;
	}

	// If the object is the original brick, don't count it
	if(%object == %brick) {
		return false;
	}

	// Make sure the brick is a valid baseplate
	if(!%object.isValidAdjacentBaseplate()) {
		return false;
	}	

	// Object is an adjacent baseplate
	return true;
}

// /**
//  * Returns whether or not the calling brick is the first brick on the server.
//  *
//  * @param  {fxDTSBrick}  %brick  The calling brick.
//  *
//  * @return {boolean}
//  */
function fxDTSBrick::isFirstBrick(%brick)
{
	// Determine the total count of existing bricks
	%brickCount = getBrickCount();

	// Determine all of the connected clients
	%count = ClientGroup.getCount();

	// Iterate through each client
	for(%i = 0; %i < %count; %i++) {

		// If any player has a ghost brick, remove it from the total count
		if(isObject(ClientGroup.getObject(%i).player.tempBrick)) {
			%brickCount--;
		}

	}

	// If this brick considers itself already planted, reduce the count by 1
	if(%brick.isPlanted) {
		%brickCount--;
	}

	// Return whether or not the brick count is zero
	return %brickCount <= 0;
}

// /**
//  * Returns whether or not the calling brick is a valid baseplate for adjacency.
//  *
//  * @param  {fxDTSBrick}  %brick  The calling brick.
//  *
//  * @return {boolean}
//  */
function fxDTSBrick::isValidAdjacentBaseplate(%brick)
{
	// If the brick has not been plated, it is not valid
	if(!%brick.isPlanted) {
		return false;
	}

	// If you are a cheeky bastard, then you may have noticed that a
	// brick in its death animation still counts as adjaceny. Even
	// though the brick can be killed later on, it's not clean.

	// If the brick is dying or dead, it is not valid
	if(%brick.isDead) {
		return false;
	}

	// If the brick is not on the ground, it is not valid
	if(!%brick.isGrounded()) {
		return false;
	}

	// If the brick is not a baseplate, it is not valid
	if(%brick.getDatablock().category !$= "Baseplates") {
		return false;
	}

	// Determine the size of the brick in grid units
	%size = %brick.getAlignmentGridSize();

	// If the brick is not wide enough to fill a grid cell, it is not valid
	if(getWord(%size, 0) == 0) {
		return false;
	}

	// If the brick is not long enough to fill a grid cell, it is not valid
	if(getWord(%size, 1) == 0) {
		return false;
	}

	// If the brick is not the correct baseplate type, it is not valid
	if(!%brick.isCorrectBaseplateType()) {
		return false;
	}

	// The brick should be valid
	return true;
}

Scripts/alignment.cs
//* Description *//
// Title: Alignment
// Author: Boom (9740)
// Defines the alignment portion of the add-on.

// /**
//  * Triggered when a brick is checked for grid alignment.
//  *
//  * @param  {fxDTSBrick}  %brick   The brick performing the alignment check.
//  *
//  * @return {boolean}
//  */
function fxDTSBrick::onAlignmentCheck(%brick)
{
	// This method would be a good spot for a package to add any
	// overrides for admins or something if you wanted to have
	// scenarios where you could ignore the alignment rules.
}

// /**
//  * Triggered when a baseplate passes an alignment check.
//  *
//  * @param  {fxDTSBrick}  %brick  The brick that passed the alignment check.
//  *
//  * @return {void}
//  */
function fxDTSBrick::onAlignmentPass(%brick)
{
	// This method purely exists for package hooks, and serves no
	// purpose in this mod's default functionality. However, an
	// external add-on may choose to add a listener to this.
}

// /**
//  * Triggered when a baseplate fails an alignment check.
//  *
//  * @param  {fxDTSBrick}  %brick  The brick that passed the alignment check.
//  *
//  * @return {void}
//  */
function fxDTSBrick::onAlignmentFail(%brick)
{
	// One quirk of alignment failure is auto alignment. When this
	// is enabled, we'll automatically align the ghost brick for
	// the player. Attempting to plant again should pass it.

	// Check for auto-align
	if($Support::Baseplate::Alignment::Auto) {
		%brick.align();
	}
}

// /**
//  * Performs an alignment check on the calling brick.
//  *
//  * @param  {fxDTSBrick}  %brick  The calling brick.
//  *
//  * @return {boolean}
//  */
function fxDTSBrick::doAlignmentCheck(%brick)
{
	// Call the alignment check to make sure we should do it
	%result = %brick.onAlignmentCheck();

	// If the result is strictly false, then do not perform the check
	if(%result $= 0) {
		return true;
	}

	// Perform the alignment check
	%success = %brick.isAligned();

	// Call the success trigger
	if(%success) {
		%brick.onAlignmentPass();
	} else {
		%brick.onAlignmentFail();
	}

	// Return whether or not the check was successful
	return %success;
}

// /**
//  * Returns whether or not the calling brick is aligned to the grid.
//  *
//  * @param  {fxDTSBrick}  %brick  The calling brick.
//  *
//  * @return {boolean}
//  */
function fxDTSBrick::isAligned(%brick)
{
	// Determine the position of the brick
	%position = getWords(%brick.getTransform(), 0, 2);

	// Determine the aligned position of the brick
	%alignment = %brick.getAlignment();

	// Return whether or not the brick is aligned
	return %position $= %alignment;
}


// /**
//  * Moves the calling brick to the nearest aligned position.
//  *
//  * @param  {fxDTSBrick}  %brick  The calling brick.
//  *
//  * @return {void}
//  */
function fxDTSBrick::align(%brick)
{
	%brick.setTransform(%brick.getAlignment());
}

// /**
//  * Returns the position of the nearest alignment for the calling brick.
//  *
//  * @param  {fxDTSBrick}  %brick  The calling brick.
//  *
//  * @return {string}
//  */
function fxDTSBrick::getAlignment(%brick)
{
	// Determine the grid attributes
	%gridCellWidth = $Support::Baseplate::Alignment::Width;
	%gridCellLength = $Support::Baseplate::Alignment::Length;
	%gridXOffset = $Support::Baseplate::Alignment::XOffset;
	%gridYOffset = $Support::Baseplate::Alignment::YOffset;

	// If baseplate alignment isn't actually enabled, then the callers of
	// this method might try to use faulty logic. We should instead use
	// the proper size of the brick, as if it was on a 1x1 cell grid.

	// Check if alignment is disabled
	if(!$Support::Baseplate::Alignment::Enabled) {

		// Use a 1x1 grid
		%gridCellWidth = 1;
		%gridCellLength = 1;

	}

	// The global position of bricks is not as neat as one would expect.
	// Every 1 meter counts as 2 blocks laterally. Because of this,
	// we need to scale down our grid to match the global size.

	// Scale the grid
	%gridCellWidth /= 2;
	%gridCellLength /= 2;
	%gridXOffset /= 2;
	%gridYOffset /= 2;

	// Determine the position of the brick
	%transform = %brick.getTransform();

	%x = getWord(%transform, 0);
	%y = getWord(%transform, 1);
	%z = getWord(%transform, 2);

	// Offset the position of the brick
	%x -= %gridXOffset;
	%y -= %gridYOffset;

	// Determine the size of the brick
	%data = %brick.getDatablock();

	%brickWidth = %data.brickSizeX;
	%brickLength = %data.brickSizeY;
	%brickHeight = %data.brickSizeZ;

	// Determine the rotation of the brick
	%rotation = %brick.angleID;

	// Determine the offset to the north west (top left) corner of the brick
	%centerXOffset = ((%rotation == 0 || %rotation == 2) ? %brickWidth : %brickLength) / 2;
	%centerYOffset = ((%rotation == 0 || %rotation == 2) ? %brickLength : %brickWidth) / 2;

	// We the center offset by two above to get from the center to the
	// corner. But again, real world distance is half that of studs.
	// To be accurate, we must half the center offset yet again.

	// Half the center offset
	%centerXOffset /= 2;
	%centerYOffset /= 2;

	// Offset the position to be at the north west (top left) corner of the brick
	%x -= %centerXOffset;
	%y -= %centerYOffset;

	// Use the bottom of the brick for vertical alignment
	// %z -= %brickHeight / 2;

	// Determine the aligned positions
	%xAlign = mFloor(%x / %gridCellWidth) * %gridCellWidth;
	%yAlign = mFloor(%y / %gridCellLength) * %gridCellLength;
	%zAlign = %z;

	// Reset the alignment to the center of the brick
	%xAlign += %centerXOffset;
	%yAlign += %centerYOffset;

	// Remove offset from the position of the brick
	%xAlign += %gridXOffset;
	%yAlign += %gridYOffset;

	// Return the alignment
	return %xAlign SPC %yAlign SPC %zAlign;
}

// /**
//  * Returns size of the calling brick in grid units.
//  *
//  * @param  {fxDTSBrick}  %brick  The calling brick.
//  *
//  * @return {float}
//  */
function fxDTSBrick::getAlignmentGridSize(%brick)
{
	// Determine the size of the brick
	%data = %brick.getDatablock();

	%sizeX = %data.brickSizeX;
	%sizeY = %data.brickSizeY;

	// Determine the rotation of the brick
	%rotation = mFloatLength(getWord(%brick.rotation, 3), 0);

	// Determine the size after rotation
	%brickWidth = mAbs(%rotation) == 90 ? %sizeY : %sizeX;
	%brickLength = mAbs(%rotation) == 90 ? %sizeX : %sizeY;

	// Determine the cell size of the grid
	%gridCellWidth = $Support::Baseplate::Alignment::Width;
	%gridCellLength = $Support::Baseplate::Alignment::Length;

	// If baseplate alignment isn't actually enabled, then the callers of
	// this method might try to use faulty logic. We should instead use
	// the proper size of the brick, as if it was on a 1x1 cell grid.

	// Check if alignment is disabled
	if(!$Support::Baseplate::Alignment::Enabled) {

		// Use a 1x1 grid
		%gridCellWidth = 1;
		%gridCellLength = 1;

	}

	// Determine the brick size in grid units
	%cellWidth = mFloor(%brickWidth / %gridCellWidth);
	%cellLength = mFloor(%brickLength / %gridCellLength);

	// Return the size of the brick in grid units
	return %cellWidth SPC %cellLength;
}

Scripts/preferences.cs
//* Description *//
// Title: Preferences
// Author: Boom (9740)
// Defines the preferences for this add-on.

// Check if RTB is enabled
if(isFile("Add-Ons/System_ReturnToBlockland/server.cs")) {

	// Make sure the preferences hook is enabled
	if(!$RTB::Hooks::ServerControl) {
		exec("Add-Ons/System_ReturnToBlockland/hooks/serverControl.cs");
	}

	// Register the RTB preferences
	RTB_registerPref("Baseplate Required", "Baseplate Rules | Baseplates", "$Support::Baseplate::Required", "bool", "Support_BaseplateRules", 1, false, false);
	RTB_registerPref("Baseplate Type", "Baseplate Rules | Baseplates", "$Support::Baseplate::Type", "list Any 1 Plate 2 Plain 3", "Support_BaseplateRules", 1, false, false);

	RTB_registerPref("Use Adjacency",     "Baseplate Rules | Adjacency", "$Support::Baseplate::Adjacency::Enabled",          "bool", "Support_BaseplateRules", 1, false, false);
	RTB_registerPref("Include Diagonals", "Baseplate Rules | Adjacency", "$Support::Baseplate::Adjacency::IncludeDiagonals", "bool", "Support_BaseplateRules", 1, false, false);

	RTB_registerPref("Use Alignment", "Baseplate Rules | Alignment", "$Support::Baseplate::Alignment::Enabled", "bool",       "Support_BaseplateRules", 1,  false, false);
	RTB_registerPref("Auto Align",    "Baseplate Rules | Alignment", "$Support::Baseplate::Alignment::Auto",    "bool",       "Support_BaseplateRules", 1,  false, false);
	RTB_registerPref("Cell Width", 	  "Baseplate Rules | Alignment", "$Support::Baseplate::Alignment::Width",   "int 1 1024", "Support_BaseplateRules", 16, false, false);
	RTB_registerPref("Cell Length",   "Baseplate Rules | Alignment", "$Support::Baseplate::Alignment::Length",  "int 1 1024", "Support_BaseplateRules", 16, false, false);
	RTB_registerPref("Grid X Offset", "Baseplate Rules | Alignment", "$Support::Baseplate::Alignment::XOffset", "int 0 1024", "Support_BaseplateRules", 0,  false, false);
	RTB_registerPref("Grid Y Offset", "Baseplate Rules | Alignment", "$Support::Baseplate::Alignment::YOffset", "int 0 1024", "Support_BaseplateRules", 0,  false, false);

}

// No preference manager is enabled
else {

	// Use Default Preference Values
	$Support::Baseplate::Required = true;
	$Support::Baseplate::Type = 1;

	$Support::Baseplate::Adjacency::Enabled = true;
	$Support::Baseplate::Adjacency::IncludeDiagonals = false;

	$Support::Baseplate::Alignment::Enabled = true;
	$Support::Baseplate::Alignment::Auto = true;
	$Support::Baseplate::Alignment::Width = 16;
	$Support::Baseplate::Alignment::Length = 16;
	$Support::Baseplate::Alignment::XOffset = 0;
	$Support::Baseplate::Alignment::YOffset = 0;

}

// Define the type enumerables system values
$Support::Baseplate::Type::Any = 1;
$Support::Baseplate::Type::Plate = 2;
$Support::Baseplate::Type::Plain = 3;
$Support::Baseplate::TypeMap = "Any 1 Plate 2 Plain 3";

// Define the distinct reason codes
$Support::Baseplate::Reason::InvalidBrick = "Invalid Brick";
$Support::Baseplate::Reason::NotWideEnough = "Not Wide Enough";
$Support::Baseplate::Reason::NotLongEnough = "Not Long Enough";
$Support::Baseplate::Reason::IncorrectType = "Incorrect Type";
$Support::Baseplate::Reason::UnalignedBaseplate = "Unaligned Baseplate";
$Support::Baseplate::Reason::NoBaseplateNeighbor = "No Baseplate Neighbor";